node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/NodeAlerter.java

```diff
@@ -0,0 +1,442 @@
+        Map<Node, AllocationResources> availableResources = constructAvailableResourcesMap(hosts, nodeChildren);
+
+        Map<Node, Integer> timesNodeCanBeRemoved = computeMaximalRepeatedRemovals(hosts, nodeChildren, availableResources);
+        return greedyHeuristicFindFailurePath(timesNodeCanBeRemoved, hosts, nodeChildren, availableResources);
+    }
+
+    private List<Node> filterTenants(List<Node> tenants, List<Node> hosts) {
+        var parentNames = hosts.stream().map(Node::hostname).collect(Collectors.toSet());
+        return tenants.stream()
+                .filter((t -> t.flavor().getType() != Flavor.Type.BARE_METAL
+                                && t.state() != Node.State.failed
+                                && t.state() != Node.State.parked))
+                .filter(t -> parentNames.contains(t.parentHostname().orElse("")))
+                .collect(Collectors.toList());
+    }
+
+    private List<Node> filterHosts(List<Node> hosts) {
+        return hosts.stream()
+                .filter(h -> h.state() != Node.State.failed && h.state() != Node.State.parked)
+                .collect(Collectors.toList());
+    }
+
+    private Optional<HostFailurePath> greedyHeuristicFindFailurePath(Map<Node, Integer> heuristic, List<Node> hosts,
+                                                   Map<Node, List<Node>> nodeChildren,
+                                                   Map<Node, AllocationResources> availableResources) {
+        if (hosts.size() == 0) return Optional.empty();
+        List<Node> parentRemovalPriorityList = heuristic.entrySet().stream()
+                .sorted(Comparator.comparingInt(Map.Entry::getValue))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toList());
+        for (int i = 1; i <= parentRemovalPriorityList.size(); i++) {
+            List<Node> hostsToRemove = parentRemovalPriorityList.subList(0, i);
+            var hostRemovalFailure = findHostRemovalFailure(hostsToRemove, hosts, nodeChildren, availableResources);
+            if (hostRemovalFailure.isPresent()) {
+                HostFailurePath failurePath = new HostFailurePath();
+                failurePath.hostsCausingFailure = hostsToRemove;
+                failurePath.failureReason = hostRemovalFailure.get();
+                return Optional.of(failurePath);
+            }
+        }
+
+        throw new IllegalStateException("No path to failure found. This should be impossible!");
+    }
+
+    int countOvercommittedHosts() {
+        List<Node> hosts = filterHosts(nodeRepository.getNodes(NodeType.host));
+        List<Node> tenants = filterTenants(nodeRepository.getNodes(NodeType.tenant), hosts);
+        var nodeMap = constructHostnameToNodeMap(hosts);
+        var nodeChildren = constructNodeChildrenMap(tenants, hosts, nodeMap);
+        var availableResources = constructAvailableResourcesMap(hosts, nodeChildren);
```

Comment:
I would have loved to, but got the impression that there's a lack of reasonable ways to return 5 variables from a function in Java. The only thing I can imagine is a class to hold pointers to the 5 different variables constructed in that preamble. Alternatively, I could make an overload which constructs the `availableResources` map from scratch since that's the only part that `countOvercommittedNodes` uses, but I can't imagine a way to reduce duplication without an increase in complexity I'm not sure if is worth it. If absolutely all duplication is forbidden I can fix it, but I'm not sure if it's worth it otherwise.
Thoughts?

Reply:
It's fine as it is then